### PR TITLE
FORT18 - Private Key Compromise

### PR DESCRIPTION
**Pull request by 0xtaf**
# FORT18 - Private Key Compromise

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [x] New Bot
- [ ] Fix a bot error
- [ ] Update a bot logic/test/documentation
- [ ] Code style update (formatting, renaming)
- [ ] Other (please describe):

## New Bot PR readiness

- [x] `npm install` command works (i.e. package-lock is updated properly)
- [x] The `name` and `description` fields of `package.json` describe the bot properly
- [x] All the libraries listed in `package.json` are being used
- [x] All the development libraries are in `devDependencies` section of `package.json`
- [x] The bot README is updated
- [x] The bot README sufficiently describes what the bot does
- [x] All the alerts that the bot emits are listed in the README
- [x] The txn/block examples listed in the README generate the expected findings
- [x] Code is easy to follow and/or properly commented
- [x] Bot contains tests
- [x] All tests pass when running `npm test`
- [x] Tests are not making real network calls (i.e. network interaction is properly mocked)
- [x] Tests cover all the paths (i.e. negative & positive test cases)
- [x] When the tests complete, no warning/errors are shown
- [x] `npm start` works without errors when executed using a jsonRpcUrl for each of the networks specified in the README
- [x] The networks specified in the README and the networks are listed in the `package.json` are the same
- [x] The code does what the README describes
- [x] Network calls are properly being cached (if duplicated calls can occur)
- [x] A performance review has been done (e.g. `Promise.all` is used whenever possible)
- [x] The findings fields are filled out correctly
- [x] Findings `protocol` is specified (if the bot is for a specific one)
- [x] `metadata` of the findings contains all the information needed.
- [x] The SDK methods are being used appropriately (e.g. `filterLog`, `filterFunction`, `getEthersProvider`)
- [x] `ABI`s match the contracts used / `args` used from `LogDescriptions` exists in the `ABI`s
- [x] The bot follows Forta recommended [best practices](https://docs.forta.network/en/latest/best-practices/)
- [x] Bot can recover from failed error calls when they can fail (e.g. errors are being caught on calls that can fail)
- [x] `repository` is added in the `package.json`

## Comments (optional)

If there is some useful information or important note for reviewing/understanding this bot please describe it here.

### CODE REVIEW

**Review by Vxatz — COMMENTED** [review]
I'd say that something we need to include in the code is a check that all the victim's assets have been drained to consider that the attack is a PK compromise. Quick thought: We could check for multiple draining-different-token transactions from the victim to the attacker in a short period


Secondary Comments:

1) The confidence level is manually set by us (hardcoded as you have it), I think `0.6` is a good choice.

2) Perhaps you can add a check for the labels at least for the 3 chains supported by the Forta API (https://github.com/NethermindEth/general-agents-module/issues/55 for more info)

3) I suggest adding a transactions counter for the testing instance in order to check the number of dropped transactions. If it's high (especially on Ethereum) we'll then need to enable sharding (and configure the persistence logic)

File: private-key-compromise/package.json [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    "info": "forta-agent info",
+    "logs": "forta-agent logs",
+    "push": "forta-agent push",
+    "disable": "forta-agent disable",
+    "enable": "forta-agent enable",
+    "keyfile": "forta-agent keyfile",
+    "test": "jest",
+    "format": "prettier --write \"src/**/*.ts\""
+  },
+  "dependencies": {
```
Comment (Vxatz) [inline private-key-compromise/package.json:36]: `bot-alert-rate` missing

File: private-key-compromise/src/.prettierrc
Comment (Vxatz) [inline private-key-compromise/src/.prettierrc:1]: Typically we put this in the root folder, although it works like that too

File: private-key-compromise/src/agent.ts [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+                if (transferObj[transfer.args.to].length > 3) {
+                  alertedAddresses.push({
+                    address: transfer.args.to,
+                    timestamp: txEvent.timestamp,
+                  });
+                  const anomalyScore = await calculateAlertRate(
+                    Number(chainId),
+                    BOT_ID,
+                    "PKC-1",
+                    ScanCountType.TransferCount
```
Comment (Vxatz) [inline private-key-compromise/src/agent.ts:128]: maybe it's better to use `.ErcTransferCount` on this

**Review by Vxatz — COMMENTED** [review]
I just remember that according to [this](https://github.com/forta-network/bot-alert-rate/blob/main/typescript/src/helpers/constants.ts#L12-L14) the `bot-alert-rate` package is not available on those 3 chains

[This](https://github.com/NethermindEth/forta-starter-kits/blob/6297a20a488422fde764d508a771c7452d3c1a7f/asset-drained/src/agent.js#L265-L266) is my implementation

File: private-key-compromise/src/.prettierrc
Comment (0xtaf) [reply private-key-compromise/src/.prettierrc:1]: I totally saw it inside root. need to see an ophthalmologist

**Review by Vxatz — APPROVED** [review]
LGTM!

**Pull request merged**